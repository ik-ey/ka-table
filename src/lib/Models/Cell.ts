export class Cell {
    public columnKey!: string;
    public rowKeyValue!: any;
}
