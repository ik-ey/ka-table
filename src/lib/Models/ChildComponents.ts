import {
  ICellEditorProps,
  ICellProps,
  ICellTextProps,
  IDataRowProps,
  IEmptyCellProps,
  IFilterRowEditorProps,
  IGroupPanelCellProps,
  IGroupPanelProps,
  IGroupRowProps,
  IGroupSummaryCellProps,
  IGroupSummaryRowProps,
  IHeadCellProps,
  IHeadCellResizeProps,
  IHeadRowProps,
  IHeaderFilterButtonProps,
  ILoadingProps,
  INoDataRowProps,
  IPagingIndexProps,
  IPagingProps,
  IPagingSizeProps,
  IPopupContentItemProps,
  IPopupContentProps,
  IRowProps,
  ISortIconProps,
  ISummaryCellProps,
  ISummaryRowProps,
  ITableBodyProps,
  ITableFootProps,
  ITableHeadProps,
} from '../props';

import { ChildComponent } from './ChildComponent';
import { ITableProps } from '../';

export class ChildComponents {
  public cell?: ChildComponent<ICellProps>;
  public cellEditor?: ChildComponent<ICellEditorProps>;
  public cellEditorInput?: ChildComponent<ICellEditorProps>;
  public cellText?: ChildComponent<ICellTextProps>;
  public dataAndDetailsRows?: ChildComponent<IRowProps>;
  public dataRow?: ChildComponent<IDataRowProps>;
  public detailsCell?: ChildComponent<IDataRowProps>;
  public detailsRow?: ChildComponent<IDataRowProps>;
  public emptyCell?: ChildComponent<IEmptyCellProps>;
  public filterRowCell?: ChildComponent<IFilterRowEditorProps>;
  public groupPanel?: ChildComponent<IGroupPanelProps>;
  public groupPanelCell?: ChildComponent<IGroupPanelCellProps>;
  public groupExpandButton?: ChildComponent<IGroupRowProps>;
  public groupCell?: ChildComponent<IGroupRowProps>;
  public groupRow?: ChildComponent<IGroupRowProps>;
  public groupSummaryRow?: ChildComponent<IGroupSummaryRowProps>;
  public groupSummaryCell?: ChildComponent<IGroupSummaryCellProps>;
  public headFilterButton?: ChildComponent<IHeaderFilterButtonProps>;
  public headCell?: ChildComponent<IHeadCellProps>;
  public headCellContent?: ChildComponent<IHeadCellProps>;
  public headCellResize?: ChildComponent<IHeadCellResizeProps>;
  public headRow?: ChildComponent<IHeadRowProps>;
  public loading?: ChildComponent<ILoadingProps>;
  public noDataRow?: ChildComponent<INoDataRowProps>;
  public noDataCell?: ChildComponent<INoDataRowProps>;
  public paging?: ChildComponent<IPagingProps>;
  public pagingIndex?: ChildComponent<IPagingIndexProps>;
  public pagingPages?: ChildComponent<IPagingProps>;
  public pagingSize?: ChildComponent<IPagingSizeProps>;
  public pagingSizes?: ChildComponent<IPagingProps>;
  public popupContent?: ChildComponent<IPopupContentProps>;
  public popupContentItem?: ChildComponent<IPopupContentItemProps>;
  public rootDiv?: ChildComponent<ITableProps>;
  public sortIcon?: ChildComponent<ISortIconProps>;
  public summaryCell?: ChildComponent<ISummaryCellProps>;
  public summaryRow?: ChildComponent<ISummaryRowProps>;
  public table?: ChildComponent<ITableProps>;
  public tableBody?: ChildComponent<ITableBodyProps>;
  public tableFoot?: ChildComponent<ITableFootProps>;
  public tableHead?: ChildComponent<ITableHeadProps>;
  public tableWrapper?: ChildComponent<ITableProps>;
}
