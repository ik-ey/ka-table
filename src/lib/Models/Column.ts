import { DataType, SortDirection } from '../enums';

import { Field } from '../types';
import { PopupPosition } from './PopupPosition';

export class Column<TData = any> {
    public colGroup?: React.ColHTMLAttributes<HTMLElement>;
    public dataType?: DataType;
    public field?: Field;
    public filter?: (value: any, filterValue: any, rowData?: any) => boolean;
    public filterRowOperator?: any;
    public filterRowValue?: any;
    public headerFilterListItems?: (props: { data?: TData[] }) => string[];
    public headerFilterPopupPosition?: PopupPosition;
    public headerFilterSearch?: (value: any, searchValue: any, rowData?: any) => boolean;
    public headerFilterSearchValue?: any;
    public headerFilterValues?: string[];
    public isDraggable?: boolean;
    public isEditable?: boolean;
    public isFilterable?: boolean;
    public isHeaderFilterPopupShown?: boolean;
    public isHeaderFilterSearchable?: boolean;
    public isResizable?: boolean;
    public isSortable?: boolean;
    public key!: string;
    public sortDirection?: SortDirection;
    public sortIndex?: number;
    public style?: React.CSSProperties;
    public title?: string;
    public visible?: boolean;
    public width?: number | string;
}
