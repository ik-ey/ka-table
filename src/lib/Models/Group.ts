
export class Group {
    public columnKey!: string;
    public enableSummary?: boolean;
}
