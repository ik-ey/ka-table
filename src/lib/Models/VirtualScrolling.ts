export class VirtualScrolling {
    public enabled?: boolean;
    public scrollTop?: number;
    public itemHeight?: ((data: any) => number) | number;
    public tbodyHeight?: number;
    public bottomInvisibleCount?: number;
    public topInvisibleCount?: number;
}
