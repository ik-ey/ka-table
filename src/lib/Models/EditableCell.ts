import { Cell } from './Cell';

export class EditableCell extends Cell {
    public editorValue?: any;
    public validationMessage?: any;
}
