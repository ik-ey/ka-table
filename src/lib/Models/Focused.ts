import { Cell } from './Cell';

export class Focused {
    public cell?: Cell;
    public cellEditorInput?: Cell;
}
