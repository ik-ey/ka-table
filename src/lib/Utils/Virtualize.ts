import { VirtualScrolling } from '../Models/VirtualScrolling';
import { checkIndexOdd } from './DataUtils';

export const isVirtualScrollingEnabled = (virtualScrolling?: VirtualScrolling) => {
    return virtualScrolling && virtualScrolling.enabled !== false;
}

export const getVirtualized = (virtualScrolling: VirtualScrolling, data: any[], isNewRowShown?: boolean, oddEvenRows?: boolean) => {
    const virtualizedData: any[] = [];
    const { scrollTop = 0, bottomInvisibleCount = 5, topInvisibleCount = 0 } = virtualScrolling;
    let { tbodyHeight = 600 } = virtualScrolling;
    let beginHeight = 0;
    let endHeight = 0;
    data.reduce((acc, value) => {
        const itemHeight = virtualScrolling.itemHeight ?
            (
                typeof virtualScrolling.itemHeight === 'number'
                    ? virtualScrolling.itemHeight
                    : virtualScrolling.itemHeight(value)
            )
            : 40;
        const topInvisibleHeight = (itemHeight * topInvisibleCount);
        if (acc >= scrollTop - itemHeight - topInvisibleHeight) {
            if (tbodyHeight + topInvisibleHeight >= -(itemHeight * bottomInvisibleCount)) {
                tbodyHeight = tbodyHeight - itemHeight;
                virtualizedData.push(value);
            } else {
                endHeight += itemHeight;
            }
        } else {
            beginHeight = acc;
            if (!isNewRowShown){
                beginHeight += itemHeight;
            }
        }
        return acc + itemHeight;
    }, 0);

    return {
        beginHeight,
        endHeight,
        virtualizedData,
        isFirstVisibleRowOdd: oddEvenRows
            ? checkIndexOdd(data?.indexOf(virtualizedData[0]))
            : undefined
    };
};
