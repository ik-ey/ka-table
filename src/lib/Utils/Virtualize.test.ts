import { getVirtualized, isVirtualScrollingEnabled } from './Virtualize';

import { VirtualScrolling } from '../Models/VirtualScrolling';

describe('Virtualize', () => {
  const data: any[] = new Array(100).fill(0).map((_, index) => index);
  describe('getVirtualized', () => {
    it('scrollTop 0', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollTop: 0,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollTop 100', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollTop: 100,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollTop 100 bottomInvisibleCount', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollTop: 100,
        tbodyHeight: 40,
        bottomInvisibleCount: 10
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });
    it('scrollTop 900 topInvisibleCount', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollTop: 900,
        tbodyHeight: 40,
        topInvisibleCount: 10
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollTop 900', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollTop: 900,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('itemHeight 40', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: 40,
        scrollTop: 0,
        tbodyHeight: 160,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('with new row', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollTop: 100,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data, true);

      expect(result).toMatchSnapshot();
    });
  });

  it('isVirtualScrollingEnabled', () => {
    expect(isVirtualScrollingEnabled()).toBeFalsy();
    expect(isVirtualScrollingEnabled({})).toBeTruthy();
    expect(isVirtualScrollingEnabled({ enabled: true })).toBeTruthy();
    expect(isVirtualScrollingEnabled({ enabled: false })).toBeFalsy();
  });
});
