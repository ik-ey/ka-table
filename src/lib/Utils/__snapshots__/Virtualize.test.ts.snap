// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virtualize getVirtualized itemHeight 40 1`] = `
Object {
  "beginHeight": 0,
  "endHeight": 3600,
  "virtualizedData": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
  ],
}
`;

exports[`Virtualize getVirtualized scrollTop 0 1`] = `
Object {
  "beginHeight": 0,
  "endHeight": 900,
  "virtualizedData": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
  ],
}
`;

exports[`Virtualize getVirtualized scrollTop 100 1`] = `
Object {
  "beginHeight": 90,
  "endHeight": 810,
  "virtualizedData": Array [
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
  ],
}
`;

exports[`Virtualize getVirtualized scrollTop 100 bottomInvisibleCount 1`] = `
Object {
  "beginHeight": 90,
  "endHeight": 760,
  "virtualizedData": Array [
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
  ],
}
`;

exports[`Virtualize getVirtualized scrollTop 900 1`] = `
Object {
  "beginHeight": 890,
  "endHeight": 10,
  "virtualizedData": Array [
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
  ],
}
`;

exports[`Virtualize getVirtualized scrollTop 900 topInvisibleCount 1`] = `
Object {
  "beginHeight": 790,
  "endHeight": 10,
  "virtualizedData": Array [
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
  ],
}
`;

exports[`Virtualize getVirtualized with new row 1`] = `
Object {
  "beginHeight": 80,
  "endHeight": 810,
  "virtualizedData": Array [
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
  ],
}
`;
